plan/refiner.go

```diff
@@ -139,36 +156,50 @@ func detachIndexScanConditions(conditions []expression.Expression, indexScan *Ph
 			indexScan.accessEqualCount = len(accessConds)
 		}
 	}
-
-	checker := &conditionChecker{
-		tableName:    indexScan.Table.Name,
-		idx:          indexScan.Index,
-		columnOffset: indexScan.accessEqualCount,
-	}
-	for _, cond := range conditions {
-		isAccess := false
-		for _, acCond := range accessConds {
-			if cond == acCond {
-				isAccess = true
+	indexScan.accessInAndEqCount = indexScan.accessEqualCount
+	// We should remove all accessConds , so that they will not be added to filter conditions.
+	for i := len(conditions) - 1; i >= 0; i-- {
+		for _, cond := range accessConds {
+			if cond == conditions[i] {
+				conditions = append(conditions[:i], conditions[i+1:]...)
 				break
 			}
 		}
-		if isAccess {
-			continue
+	}
+	for curIndex := indexScan.accessEqualCount; curIndex < len(indexScan.Index.Columns); curIndex++ {
+		checker := &conditionChecker{
+			tableName:    indexScan.Table.Name,
+			idx:          indexScan.Index,
+			columnOffset: curIndex,
 		}
-		cond = pushDownNot(cond, false)
-		if indexScan.accessEqualCount >= len(indexScan.Index.Columns) ||
-			!checker.check(cond) {
-			filterConds = append(filterConds, cond)
+		// First of all we should extract all of in/eq expressions. We still need to consider eq functions,
+		// because they may not be counted before. e.g. a = 1 and b in (3,4) and c = 5. Only a = 1 will be counted
+		// in accessEqualCount.
+		accessIdx := checker.extractEqOrInFunc(conditions)
+		if accessIdx != -1 {
+			indexScan.accessInAndEqCount++
+			accessConds = append(accessConds, conditions[accessIdx])
+			if indexScan.Index.Columns[curIndex].Length != types.UnspecifiedLength {
+				filterConds = append(filterConds, conditions[accessIdx])
+			}
+			conditions = append(conditions[:accessIdx], conditions[accessIdx+1:]...)
 			continue
 		}
-		accessConds = append(accessConds, cond)
-		if indexScan.Index.Columns[indexScan.accessEqualCount].Length != types.UnspecifiedLength ||
-			// TODO: it will lead to repeated compution cost.
-			checker.shouldReserve {
-			filterConds = append(filterConds, cond)
-			checker.shouldReserve = false
+		// If we fail to find any in or eq expression, we should consider all of other conditions for the next column.
+		for _, cond := range conditions {
```

Comment:
Extract this block to a function and put it in `if accessIdx == -1`, then we don't need to call `continue`.